### ISSUE TITLE
Ruby Realization

### ISSUE THREAD

**Issue by feddischson** [feddischson/include_gardener#17]
One of the goal with v1.0.0 was to re-write the architecture to enable multiple languages.

For this reason, it would be interesting to write a Solver_Rb class, derived from the Solver class, which enables the detection of Ruby `require` statements in Ruby code.

Please leave a comment if somebody is interested in writing this extension.

**Comment by bstee615** [comment]
@feddischson I'd like to work on this.

**Comment by bstee615** [comment]
@feddischson I got the build system together with CMake, but I'm getting some errors, even with `std::` type things. Pasted some examples below. Attempting to build on Visual Studio 15, Windows 10. I will try on Linux. Any advice? What OS are you building on currently?

```
1>C:\code\oss\include_gardener\src\statement_detector.cpp(160): error C2664: 'std::basic_ifstream<char,std::char_traits<char>>::basic_ifstream(const std::basic_ifstream<char,std::char_traits<char>> &)': cannot convert argument 1 from 'int' to 'const std::experimental::filesystem::v1::path &'
1>C:\code\oss\include_gardener\src\statement_detector.cpp(160): note: Reason: cannot convert from 'int' to 'const std::experimental::filesystem::v1::path'
1>C:\code\oss\include_gardener\src\statement_detector.cpp(160): note: use of undefined type 'std::experimental::filesystem::v1::path'
1>C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\include\fstream(35): note: see declaration of 'std::experimental::filesystem::v1::path'
1>C:\code\oss\include_gardener\src\statement_detector.cpp(161): error C2664: 'void INCLUDE_GARDENER::Statement_Detector::process_stream(std::istream &,const std::string &)': cannot convert argument 2 from 'int' to 'const std::string &'
1>C:\code\oss\include_gardener\src\statement_detector.cpp(161): note: Reason: cannot convert from 'int' to 'const std::string'
1>C:\code\oss\include_gardener\src\statement_detector.cpp(161): note: No constructor could take the source type, or constructor overload resolution was ambiguous
1>C:\code\oss\include_gardener\src\statement_detector.cpp(173): error C2027: use of undefined type 'std::deque'
```

**Comment by feddischson** [comment]
Sounds good. I've created a very small overview [here](http://www.easyjet.com/de/Booking/Receipt?ref=EWQZJ8S&k=gAULWo2aIO2mG%2BVRv48kI4lGbec3u8qFOrxyuCNEtiw%3D). Feel free to extend it if something is missing.

I am using Linux and I tested the build with clang and gcc.

**Comment by bstee615** [comment]
Yes, I found the Ruby doc after reading the comments on #16 and I will definitely use them. Still figuring our how $LOAD_PATH works in Ruby as I have not written in Ruby before. I'm planning on first supporting `require_relative` in early commits as it is the easiest, then expanding functionality to `load` and `require`. Will try to update as I go along.

**Comment by feddischson** [comment]
I think it is not required to support `$LOAD_PATH`. There is also no plan to support such dynamic things in Python.

**Comment by bstee615** [comment]
Ok. I've finished `require_relative` and am working on implementing `require` and `load` without `$LOAD_PATH`. If we just execute a shell command from the solver at startup, we can grab the `$LOAD_PATH` pretty easily. However, of course, it would require the user to have Ruby installed. Should I implement this also, or should we stay away from that kind of stuff?

**Comment by feddischson** [comment]
I see two possibilities:
 - Adding a command-line option which says, if `$LOAD_PATH` is extracted via ruby call
 - Adding a command-line option which specifies the `$LOAD_PATH` without executing ruby

Both possibilities could be realized in parallel.
In my opinion, the second option would be a must-have feature. A user could define exactly which paths should be considered. 

The first option is more a nice-to-have feature. The following scenario would then be interesting: A user could use both options together. The second option to extend the first one.

What do you think?

**Comment by bstee615** [comment]
I agree that the `$LOAD_PATH` should be specified somehow as otherwise, `require` would be quite useless. However, it seems that even in the default Ruby setup `$LOAD_PATH` is quire long, with ten or so absolute paths, and this would not be good to type out.

I figure we can have an option to put in a few paths by command line and have `include_gardener` try to execute Ruby to get the rest of it.
